### ISSUE TITLE
Implement a better interface for DataPostprocessor

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#391]
DataPostprocessor currently declares virtual functions like this:

```
  virtual
  void
  compute_derived_quantities_scalar (const std::vector<double>         &uh,
                                     const std::vector<Tensor<1,dim> > &duh,
                                     const std::vector<Tensor<2,dim> > &dduh,
                                     const std::vector<Point<dim> >    &normals,
                                     const std::vector<Point<dim> >    &evaluation_points,
                                     std::vector<Vector<double> >      &computed_quantities) const;
```

The problem is that if we ever decide that we want to pass more information (e.g., the `material_id` of the cell on which we are currently evaluating, so that a user program can look up cell information), then this doesn't work: we can't just change the signature of the function to pass an additional value, since that would make user programs incompatible. In the past, we added a new virtual function with an extra argument whose default implementation just called the old function with the additional argument dropped. If a user overloads the old function, then it is called; if the user overloads the new function, then that works as well.

Clearly, this does not scale and we will be in trouble every time we want to add an argument. The way I envision this can be done in a reasonably extensible way is to declare them as follows:

```
  struct ScalarInputs {
      const std::vector<double>         uh,
      const std::vector<Tensor<1,dim> > duh,
      const std::vector<Tensor<2,dim> > dduh,
      const std::vector<Point<dim> >    normals,
      const std::vector<Point<dim> >    evaluation_points,
  };

  virtual
  void
  compute_derived_quantities_scalar (const ScalarInputs &scalar_inputs,
                                     std::vector<Vector<double> >      &computed_quantities) const;
```

Using such a scheme, we could add new fields to the `ScalarInputs` structure without breaking old programs -- they'd simply ignore the additional field, but it wouldn't affect the signature of the function and consequently the ability to overload.

**Comment by bangerth** [comment]
The issue alluded above with `Tensor` vs `Point` for normal vectors is related to #496.

**Comment by bangerth** [comment]
I wrote the following for a colleague about design mistakes. I think it would be useful to copy-paste here, as an example we can point to in the future:
...................

An example of a design mistake:
deal.II has a class called DataOut that is used to take a solution field (characterized by a Triangulation for the geometry of cells, a DoFHandler to identify which degrees of freedom live where, and a solution vector for the values of these degrees of freedom) and turn that into graphical output that can then be read, for example, by Visit/Paraview/gnuplot/...

There is a separate class called DataPostprocessor that can interact with DataOut in such a way that it is possible to not only output the *solution*, but instead *derived quantities* involving the solution and its gradients. This is useful, for example, in compressible fluid flow simulators where the solution vector contains the mass density, momentum density, and energy density, but you are interested in the temperature (which can be computed from the energy density and the mass density, as well as an equation of state).

DataPostprocessor was initially written as a class with an abstract virtual function that receives the values, gradients, and Hessians of the solution. Later, we added the location of the point at which the solution is evaluated. (In reality, the function gets vectors of these quantities as it is called for all support points on the same cell at once, but that's not important here.) Then we discovered that people wanted also a pointer to the cell that's being evaluated right now, the material id, the subdomain id, and many other things.

These are not easy to add because it would require to add arguments to an abstract virtual function, and call the new function from DataOut. If you do this, you break all user codes because they only implement the *old* interface. So it's very difficult to undo such a design decision in a backward compatible way; it's also difficult to agree which arguments exactly this function should get.

We had known about this poor design for many years (I would say at least since around 2008 or so) but the difficulty of finding a useful solution, and in particular the pain involved in breaking user codes made us hesitant to actually fix it. I opened a bug report about this much later, when I started to have an idea how one can address this:
  https://github.com/dealii/dealii/issues/391
It was finally fixed in
  https://github.com/dealii/dealii/issues/3333
but it turned out that it's actually quite difficult to pass the pointer to the cell as I had hoped for because the DataPostprocessor can be used from many contexts:
  https://github.com/dealii/dealii/issues/3512
It took a substantial additional effort to fix this as well:
  https://github.com/dealii/dealii/pull/3689

I think this is probably the best example of poor design I've come across recently. We used to have many more examples, but they are often easier to fix because one can deprecate a function with a poor interface, add a replacement somewhere that is being called by the deprecated interface, and a couple of years later remove the deprecated function. All of this is greatly complicated if the poor design is a pure virtual function interface, as the example above shows.
